crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1187 @@
+    ctx->expected_sender = NULL;
+    if (name == NULL)
+        return 1;
+    return (ctx->expected_sender = X509_NAME_dup(name)) != NULL;
+}
+
+/*
+ * Set the X509 name of the issuer. Set in the PKIHeader.
+ * Returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_issuer(OSSL_CMP_CTX *ctx, const X509_NAME *name)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    X509_NAME_free(ctx->issuer);
+    ctx->issuer = NULL;
+    if (name == NULL)
+        return 1;
+    return (ctx->issuer = X509_NAME_dup(name)) != NULL;
+}
+
+/*
+ * Set the subject name that will be placed in the certificate
+ * request. This will be the subject name on the received certificate.
+ * Returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_subjectName(OSSL_CMP_CTX *ctx, const X509_NAME *name)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    X509_NAME_free(ctx->subjectName);
+    ctx->subjectName = NULL;
+    if (name == NULL)
+        return 1;
+    return (ctx->subjectName = X509_NAME_dup(name)) != NULL;
+}
+
+/*
+ * Set the X.509v3 certificate request extensions to be used in IR/CR/KUR.
+ * Returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set0_reqExtensions(OSSL_CMP_CTX *ctx, X509_EXTENSIONS *exts)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (sk_GENERAL_NAME_num(ctx->subjectAltNames) > 0 && exts != NULL
+            && X509v3_get_ext_by_NID(exts, NID_subject_alt_name, -1) >= 0) {
+        CMPerr(0, CMP_R_MULTIPLE_SAN_SOURCES);
+        return 0;
+    }
+    sk_X509_EXTENSION_pop_free(ctx->reqExtensions, X509_EXTENSION_free);
+    ctx->reqExtensions = exts;
+    return 1;
+}
+
+/* returns 1 if ctx contains a Subject Alternative Name extension, else 0 */
+int OSSL_CMP_CTX_reqExtensions_have_SAN(OSSL_CMP_CTX *ctx)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return -1;
+    }
+    /* if one of the following conditions 'fail' this is not an error */
+    return ctx->reqExtensions != NULL
+        && X509v3_get_ext_by_NID(ctx->reqExtensions,
+                                 NID_subject_alt_name, -1) >= 0;
+}
+
+/*
+ * Add a GENERAL_NAME structure that will be added to the CRMF
+ * request's extensions field to request subject alternative names.
+ * Returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_push1_subjectAltName(OSSL_CMP_CTX *ctx,
+                                      const GENERAL_NAME *name)
+{
+    if (ctx == NULL || name == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    if (OSSL_CMP_CTX_reqExtensions_have_SAN(ctx) == 1) {
+        CMPerr(0, CMP_R_MULTIPLE_SAN_SOURCES);
+        return 0;
+    }
+
+    if (ctx->subjectAltNames == NULL
+            && (ctx->subjectAltNames = sk_GENERAL_NAME_new_null()) == NULL)
+        return 0;
+    return sk_GENERAL_NAME_push(ctx->subjectAltNames, GENERAL_NAME_dup(name));
```

Comment:
how is this going to work, what if GN_dup works, but sk_GN_push fails ?

Reply:
Fixed.